From 09ed94391946d1cc5d3d27d9513de24eea6d992d Mon Sep 17 00:00:00 2001
From: Ilan Peer <ilan.peer@intel.com>
Date: Tue, 14 Apr 2026 12:04:25 +0530
Subject: [PATCH 1/2] nl80211: Don't try to scan on AP interface with active
 links

When the kernel driver doesn't support scanning on an AP interface,
nl80211 drivers attempts to switch to station interface to perform a
scan. However, when the AP is affiliated with an AP MLD and there are
already active links this is not valid. In such a case, fail the scan.

https://git.w1.fi/cgit/hostap/commit/?id=c7930eae3b1ceab5fce9c30a3e46c97f4c72938e

Signed-off-by: Ilan Peer <ilan.peer@intel.com>
Reviewed-by: Andrei Otcheretianski <andrei.otcheretianski@intel.com>
Signed-off-by: Benjamin Berg <benjamin.berg@intel.com>
Signed-off-by: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
---
 src/drivers/driver_nl80211_scan.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/drivers/driver_nl80211_scan.c b/src/drivers/driver_nl80211_scan.c
index 8d7f0921aee3..b661a9f74524 100644
--- a/src/drivers/driver_nl80211_scan.c
+++ b/src/drivers/driver_nl80211_scan.c
@@ -459,8 +459,16 @@ int wpa_driver_nl80211_scan(struct i802_bss *bss,
 
 			/*
 			 * mac80211 does not allow scan requests in AP mode, so
-			 * try to do this in station mode.
+			 * try to do this in station mode. Do so only if there
+			 * are not active links in the AP MLD.
 			 */
+			wpa_printf(MSG_DEBUG,
+				   "nl80211: AP scan failed on %s. Links=0x%x",
+				   bss->ifname, bss->valid_links);
+
+			if (bss->valid_links)
+				goto fail;
+
 			drv->ap_scan_as_station = drv->nlmode;
 			if (wpa_driver_nl80211_set_mode(
 				    bss, NL80211_IFTYPE_STATION) ||
-- 
2.34.1

